go/vumitools/middleware.py

```diff
@@ -603,6 +604,151 @@ def handle_failure(self, failure, connector_name):
+        self.op_mode = self.config.op_mode
+        if self.op_mode not in self.KNOWN_MODES:
+            raise ConfigError('Unknown op_mode: %s' % (
+                self.op_mode,))
+        self.metric_connectors_specified = (
+            self.config.metric_connectors is not None)
+        self.metric_connectors = set(self.config.metric_connectors or [])
+
+    @inlineCallbacks
+    def setup_middleware(self):
+        self.validate_config()
+        self.metric_publisher = yield self.worker.start_publisher(
+            MetricPublisher)
+        # We don't use a VumiApi here because we don't have a Riak config for
+        # it.
+        self.redis = yield TxRedisManager.from_config(
+            self.config.redis_manager)
+        self.metric_manager = MetricManager(
+            self.manager_name + '.', publisher=self.metric_publisher)
+        self.metric_manager.start_polling()
+
+    def teardown_middleware(self):
+        self.metric_manager.stop_polling()
+        return self.redis.close_manager()
+
+    def get_or_create_metric(self, name, metric_class, *args, **kwargs):
+        """
+        Get the metric for `name`, create it with
+        `metric_class(*args, **kwargs)` if it doesn't exist yet.
+        """
+        if name not in self.metric_manager:
+            self.metric_manager.register(metric_class(name, *args, **kwargs))
+        return self.metric_manager[name]
+
+    def get_counter_metric(self, name):
+        return self.get_or_create_metric(name, Count)
+
+    def increment_counter(self, prefix, message_type):
+        metric = self.get_counter_metric('%s.%s' % (prefix, message_type))
+        metric.inc()
+
+    def get_acc_key(self, msg):
+        mdh = MessageMetadataHelper(
+            self.vumi_api, msg, conversation_cache=self._conversation_cache)
+        acc_key = yield mdh.get_account_key()
+        returnValue(acc_key)
+
+    def get_conv_key(self, msg):
+        mdh = MessageMetadataHelper(
+            self.vumi_api, msg, conversation_cache=self._conversation_cache)
+        conv_key = yield mdh.get_conversation_key()
+        returnValue(conv_key)
+
+    def get_prefix(self, msg):
+        acc_key = self.get_acc_key(msg)
+        conv_key = self.get_conv_key(msg)
+        prefix = "."+acc_key+".conversations."+conv_key+"."
+        returnValue(prefix)
+
+    def fire_inbound_metrics(self, prefix, msg):
+        self.increment_counter(prefix, 'messages_received')
+
+    def fire_inbound_conversation_metrics(self, name, msg):
+        self.fire_inbound_metrics(name, msg)
+
+    def fire_outbound_metrics(self, prefix, msg):
+        self.increment_counter(prefix, 'messages_sent')
+
+    def fire_outbound_conversation_metrics(self, name, msg):
+        self.fire_outbound_metrics(name, msg)
+
+    @inlineCallbacks
+    def handle_inbound(self, message, connector_name):
+        name = self.get_prefix(message)
+        self.fire_inbound_metrics(name, message)
+        returnValue(message)
```

Comment:
One needs to `yield` in an `inlineCallbacks` method (otherwise the function isn't a generator, and `inlineCallbacks` only makes sense with generators).